src/details/ArborX_DetailsDistributor.hpp

```diff
@@ -31,6 +31,103 @@ namespace ArborX
+  ARBORX_ASSERT(counts.empty());
+  ARBORX_ASSERT(permutation_indices.extent_int(0) == 0);
+  ARBORX_ASSERT(batched_ranks.size() + 1 == batched_offsets.size());
+  static_assert(
+      std::is_same<typename InputView::non_const_value_type, int>::value, "");
+  static_assert(std::is_same<typename OutputView::value_type, int>::value, "");
+
+  // In case all the batches are empty, return an empty list of unique_ranks and
+  // counts, but still have one element in offsets. This is conforming with
+  // creating the total offsets from batched_ranks and batched_offsets ignoring
+  // empty batches and calling sortAndDetermeineBufferLayout.
+  offsets.push_back(0);
+
+  auto const n = batched_ranks.size();
+  if (n == 0 || lastElement(batched_offsets) == 0)
+    return;
+
+  using DeviceType = typename InputView::traits::device_type;
+  using ExecutionSpace = typename InputView::traits::execution_space;
+
+  // Find the indices in batched_ranks for which the rank changes and store
+  // these ranks and the corresponding offsets in a new container that we can be
+  // sure to be large enough.
+  Kokkos::View<int *, DeviceType> compact_offsets(
+      Kokkos::ViewAllocateWithoutInitializing(batched_offsets.label()),
+      batched_offsets.size());
+  Kokkos::View<int *, DeviceType> compact_ranks(
+      Kokkos::ViewAllocateWithoutInitializing(batched_ranks.label()),
+      batched_ranks.size());
+
+  int n_unique_ranks;
+  Kokkos::parallel_scan(
+      ARBORX_MARK_REGION("compact_offsets_and_ranks"),
+      Kokkos::RangePolicy<ExecutionSpace>(0, n),
+      KOKKOS_LAMBDA(unsigned int i, int &update, bool last_pass) {
+        if ((i < batched_ranks.size() - 1 &&
+             batched_ranks(i + 1) != batched_ranks(i)) ||
+            i == batched_ranks.size() - 1)
+        {
+          if (last_pass)
+          {
+            compact_ranks(update) = batched_ranks(i);
+            compact_offsets(update + 1) = batched_offsets(i + 1);
+          }
+          ++update;
+        }
+      },
+      n_unique_ranks);
+
+  // Now create subviews containing the elements we actually need, copy
+  // everything to the CPU and fill the output variables.
+  auto restricted_offsets = InputView(
+      compact_offsets, std::make_pair(1, static_cast<int>(n_unique_ranks + 1)));
+  auto restricted_unique_ranks = InputView(
+      compact_ranks, std::make_pair(0, static_cast<int>(n_unique_ranks)));
+
+  auto const unique_ranks_host = Kokkos::create_mirror_view_and_copy(
+      Kokkos::HostSpace(), restricted_unique_ranks);
+  unique_ranks.reserve(unique_ranks_host.size());
+  auto const offsets_host = Kokkos::create_mirror_view_and_copy(
+      Kokkos::HostSpace(), restricted_offsets);
+  offsets.reserve(restricted_offsets.size() + 1);
+  counts.reserve(restricted_offsets.size());
+
+  for (unsigned int i = 0; i < restricted_offsets.size(); ++i)
+  {
+    int const count =
+        i == 0 ? offsets_host(0) : offsets_host(i) - offsets_host(i - 1);
```

Comment:
This is really hard to parse. Would prefer getting rid of subviews the `for` from ` on, always doing `count = offsets_host(i) - offsets_host(i-1);`

Reply:
Can you clarify what the last sentence should read?

Reply:
I was trying to say that you don't need to copy a subset of elements from device to host. Just copy all of them, and then use proper indexing. Or I may be missing something.